src/System.Private.ServiceModel/src/System/IdentityModel/Tokens/X509AsymmetricSecurityKey.cs

```diff
@@ -0,0 +1,450 @@
+            if (algorithmObject != null)
+            {
+                SignatureDescription description = algorithmObject as SignatureDescription;
+                if (description != null)
+                    return description.CreateFormatter(privateKey);
+
+                try
+                {
+                    AsymmetricSignatureFormatter asymmetricSignatureFormatter = algorithmObject as AsymmetricSignatureFormatter;
+                    if (asymmetricSignatureFormatter != null)
+                    {
+                        asymmetricSignatureFormatter.SetKey(privateKey);
+                        return asymmetricSignatureFormatter;
+                    }
+                }
+                catch (InvalidCastException e)
+                {
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.AlgorithmAndPrivateKeyMisMatch, e));
+                }
+
+                throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new CryptographicException(SR.Format(SR.UnsupportedAlgorithmForCryptoOperation,
+                       algorithm, nameof(GetSignatureFormatter))));
+            }
+
+            switch (algorithm)
+            {
+                case SignedXml.XmlDsigDSAUrl:
+
+                    // Ensure that we have a DSA algorithm object.
+                    DSA dsa = (this.PrivateKey as DSA);
+                    if (dsa == null)
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.PrivateKeyNotDSA));
+#pragma warning disable CA5351 // Do Not Use Broken Cryptographic Algorithms
+                    return new DSASignatureFormatter(dsa);
+#pragma warning restore CA5351 // Do Not Use Broken Cryptographic Algorithms
+
+                case SignedXml.XmlDsigRSASHA1Url:
+                    // Ensure that we have an RSA algorithm object.
+                    RSA rsa = (this.PrivateKey as RSA);
+                    if (rsa == null)
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.PrivateKeyNotRSA));
+                    return new RSAPKCS1SignatureFormatter(rsa);
+
+                case SecurityAlgorithms.RsaSha256Signature:
+                    // Ensure that we have an RSA algorithm object.
+                    RSA rsaSha256 = (privateKey as RSA);
+                    if (rsaSha256 == null)
+                        throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.PrivateKeyNotRSA));
+                    return new RSAPKCS1SignatureFormatter(rsaSha256);
+
+                default:
+                    throw DiagnosticUtility.ExceptionUtility.ThrowHelperError(new NotSupportedException(SR.Format(SR.UnsupportedCryptoAlgorithm, algorithm)));
+            }
+        }
+
+        private static AsymmetricAlgorithm LevelUpRsa(AsymmetricAlgorithm asymmetricAlgorithm, string algorithm)
```

Comment:
If I understand correctly, this code can't ever do anything, because you have a copy of it in the PrivateKey property getter; and the only thing you ever pass in is the PrivateKey property value.

Reply:
You are right. This is a bug brought over from .NET Framework. I'll remove this method.

---
In reply to: [274025148](https://github.com/dotnet/wcf/pull/3538#discussion_r274025148) [](ancestors = 274025148)